torch/_inductor/compile_fx.py

```diff
@@ -174,11 +174,11 @@ def compile_fx_inner(
 
 
 def clone_preserve_strides(x):
-    needed_size = (
-        sum((shape - 1) * stride for shape, stride in zip(x.size(), x.stride())) + 1
-    )
-    buffer = torch.as_strided(x, (needed_size,), (1,)).clone()
-    return torch.as_strided(buffer, x.size(), x.stride())
+    # This is *bad*, x.storage() is deprecated and emits warnings
+    # TODO: figure out a way to get "# elements in the storage, in the current dtype of x"
+    needed_size = x.storage().size()
+    buffer = torch.as_strided(x, (needed_size,), (1,), 0).clone()
+    return torch.as_strided(buffer, x.size(), x.stride(), x.storage_offset())
```

Comment:
@ezyang this is just a test that fixed a bunch of CI failures locally for me.

The problem is that for code like this:
```
@torch.compile
def f(x):
    return x.as_strided(shape, stride)

base = torch.ones(20)
slice = base[5:15]
out = f(base)
```

Inductor has some logic to clone the inputs, but it doesn't respect the storage offset of said inputs. The above example breaks, because we will put `torch.ops.aten.as_strided(x, size, stride)` into the graph, and the storage_offset will be inferred from the input, to (incorrectly) be 0 instead of 5.

This used to work before the current refactor, because aot autograd was not passing output-that-view-input ops into the graph at all, and instead, it would just dump the (hardcoded) sizes/strides of views as outputs into the graph. With the changes in the aot autograd PR, we're now trusting inductor to return us an output alias with the correct metadata. Which... we kinda need to, since we talked a few weeks ago about how inductor can choose its own output memory format, so we need inductor to tell us what the strides of the outputs are in general.

Reply:
Although as I mentioned in the comment, I'm not sure that this is the best approach:

(1) I'm using the deprecated storage API. When I try `x._storage().size()` instead, I get the wrong thing: I get "number of elements in the storage assuming each element is 1 byte". What I really want is "number of elements in the storage, given that the storage is the same dtype as `x`"

(2) This will break XLA, although we can work around it. (it will also not work for sparse tensors, etc, but i'm not really sure that the original impl worked well with sparse tensors either, given the as strided calls).

Reply:
Explain to me why compositional as strided (in this case, you only need compositional storage offset: the storage offset you apply is relative to the preexisting offset from the input) doesn't work first

Reply:
Hmm so... it doesn't seem to me like compositional as_strided/storage_offset is enough here.

* If the original input tensor had storage_offset=5, that's because it was some view of a larger storage
* If we want the output to also properly preserve storage_offset=5, then the output needs to also be a view off of the same size storage as the input
* So we actually do need to clone the entire input storage, before taking an equivalent view off of it (which is what the snippet is trying to do)